http/dotnet/httpclient/src/Middleware/RetryHandler.cs

```diff
@@ -0,0 +1,185 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Kiota.Http.HttpClient.Extensions;
+using Microsoft.Kiota.Http.HttpClient.Middleware.Options;
+
+namespace Microsoft.Kiota.Http.HttpClient.Middleware
+{
+    /// <summary>
+    /// A <see cref="DelegatingHandler"/> implementation using standard .NET libraries.
+    /// </summary>
+    public class RetryHandler : DelegatingHandler
+    {
+        private const string RetryAfter = "Retry-After";
+        private const string RetryAttempt = "Retry-Attempt";
+        private double _mPow = 1;
+
+        /// <summary>
+        /// RetryOption property
+        /// </summary>
+        internal RetryHandlerOption RetryOption
+        {
+            get; set;
+        }
+
+        /// <summary>
+        /// Construct a new <see cref="RetryHandler"/>
+        /// </summary>
+        /// <param name="retryOption">An OPTIONAL <see cref="RetryHandlerOption"/> to configure <see cref="RetryHandler"/></param>
+        public RetryHandler(RetryHandlerOption retryOption = null)
+        {
+            RetryOption = retryOption ?? new RetryHandlerOption();
+        }
+
+        /// <summary>
+        /// Send a HTTP request 
+        /// </summary>
+        /// <param name="httpRequest">The HTTP request<see cref="HttpRequestMessage"/>needs to be sent.</param>
+        /// <param name="cancellationToken">The <see cref="CancellationToken"/> for the request.</param>
+        /// <returns></returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage httpRequest, CancellationToken cancellationToken)
+        {
+            RetryOption = httpRequest.GetMiddlewareOption<RetryHandlerOption>() ?? RetryOption;
```

Comment:
defensive programing on the request

Reply:
https://github.com/microsoft/kiota/pull/589/commits/7f9ba26c372017e7c79b3dda0a0661000e0d120b